packages/server/src/utils/workflow.ts

```diff
@@ -180,16 +183,44 @@ function transformWorkflowEvent(
   if (event instanceof AgentToolCallResult) {
     const rawOutput = event.data.raw;
 
+    const isValidToolOutput = rawOutput && typeof rawOutput === "object";
+    if (!isValidToolOutput) return event; // skip transform if valid tool output is not found
+
     // if AgentToolCallResult contains sourceNodes, convert it to SourceEvent
     if (
-      rawOutput &&
-      typeof rawOutput === "object" &&
       "sourceNodes" in rawOutput // TODO: better use Zod to validate and extract sourceNodes from toolCallResult
     ) {
       return toSourceEvent(
         rawOutput.sourceNodes as unknown as NodeWithScore<Metadata>[],
       );
     }
+
+    // if AgentToolCallResult contains artifact, it's output from codeGenerator tool, convert it to ArtifactEvent with code data
+    if (
+      "artifact" in rawOutput // TODO: better use Zod to validate and extract artifact from toolCallResult
+    ) {
+      const { file_path, code } = rawOutput.artifact as {
+        file_path: string;
+        code: string;
+      };
+
+      const filename = file_path.split("/").pop();
+      const language = filename?.split(".").pop();
+
+      // TODO: make it able to streaming code and document content?
+      const codeArtifact: CodeArtifact = {
+        type: "code", // TODO: handle other types of artifacts (e.g. document)
+        version: 1, // TODO: handle version
+        currentVersion: true,
+        data: {
+          file_name: filename,
+          code,
+          language,
+        } as CodeArtifactData,
+      };
+
+      return new ArtifactEvent({ type: "artifact", data: codeArtifact });
```

Comment:
that's a good idea, if i understand this correctly, if we're using a normal workfow (not agent workflow) and we want an artifact, we could emit a ArtifactEvent when we are creating an artifact (e.g. a report generator workflow might want to generate a document)